rfcs/0010-mempool-dos-mitigation.md

```diff
@@ -0,0 +1,302 @@
+transaction. Where possible, we prefer to charge for things rather than banning
+them. We don't mind if someone uses a lot of resources, so long as those
+resources are paid for at a price that network participants would be happy with.
+
+### When we first recieve a transaction: rules
+
+When we receive a gossiped transaction, we will check the below constraints. If
+any of the checks fail, we ignore the transaction and do not gossip it.
+
+* Validity:
+  1. The signature.
+  2. The sender's account exists.
+  3. The sender's balance (inclusive of any pending txs from the same account
+  with lower nonce) is >= the transaction amount + fee.
+  4. The tx nonce is equal to the sending account's next nonce, inclusive of
+  transactions already in the mempool. (If it conflicts with one in the mempool,
+  see below.)
+* Mining probability:
+  1. The (adjusted?) fee is >= the lowest fee paid in a transaction included in
+     the last `min_fee_window` blocks, times `min_fee_discount_factor`, a value
+     in the interval (0,1). FIXME: This minimum is attacker controlled. An
+     attacker who wins at least one slot and proves one SNARK can set it to
+     zero. Can't use a (exponentially weighted) moving average either, for the
+     opposite issue. Maybe don't do a set minimum and rely on the fee rules for
+     when the pool is full?
+
+If we get any transactions where the signature is invalid, we blacklist the
+sending node immediately. No honest node sends transactions like that. If we get
+transactions that fail for other reasons - the ones that depend on the current
+chain state - then we blacklist the node after `bad_tx_blacklist_limit` bad
+transactions (FIXME use pre-existing banlist and scores?). Honest nodes that are
+out of date might innocently send us transactions that are not valid, but they
+won't send us a lot of them. If a node detects it is substantially behind the
+network, it should disable gossip until it catches up.
+
+#### Replacing transactions
+
+A useful feature of existing cryptocurrencies is the ability to replace a
+pending transaction with a different one by broadcasting a new one with the same
+nonce. Users can e.g. cancel payments by replacing them with a no-op transaction
+(send themselves $0) or resend them with a higher fee if they're processing too
+slow.
+
+We want to have this feature, but it allows an attacker to make proposers
+process transactions which won't eventually get mined (the ones that are
+replaced), which violates our first guiding principle. So we require an added
+fee, `mempool_replace_fee`, representing the cost to the network to process and
+store a pending transaction, and add a field to transaction payloads counting
+the number of replacements that occurred for a given nonce. A transaction's
+total fee is its original transaction fee plus the replacement count times
+`mempool_replace_fee`. If a node receives a new transaction with the same nonce
```

Comment:
should different kinds of replacements be distinguished, like cancellation vs. increased fee?